hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-router/src/test/java/org/apache/hadoop/yarn/server/router/TestRouterAuditLogger.java

```diff
@@ -143,18 +168,75 @@ private void testFailureLogFormatHelper(ApplicationId appId,
    * Test the AuditLog format for failure events with the various
    * parameters.
    */
-  private void testFailureLogFormat() {
-    testFailureLogFormatHelper(null, null);
-    testFailureLogFormatHelper(APPID, null);
-    testFailureLogFormatHelper(null, SUBCLUSTERID);
-    testFailureLogFormatHelper(APPID, SUBCLUSTERID);
+  private void testFailureLogFormat(boolean checkIP) {
+    testFailureLogFormatHelper(checkIP, null, null);
+    testFailureLogFormatHelper(checkIP, APPID, null);
+    testFailureLogFormatHelper(checkIP, null, SUBCLUSTERID);
+    testFailureLogFormatHelper(checkIP, APPID, SUBCLUSTERID);
   }
 
   /**
    *  Test {@link RouterAuditLogger}.
    */
-  @Test public void testRouterAuditLogger() throws Exception {
-    testSuccessLogFormat();
-    testFailureLogFormat();
+  @Test
+  public void testRouterAuditLoggerWithOutIP() throws Exception {
+    testSuccessLogFormat(false);
+    testFailureLogFormat(false);
+  }
+
+  /**
+   * A special extension of {@link TestRPC.TestImpl} RPC server with
+   * {@link TestRPC.TestImpl#ping()} testing the audit logs.
+   */
+  private class MyTestRouterRPCServer extends TestRpcBase.PBServerImpl {
+    @Override
+    public TestProtos.EmptyResponseProto ping(
+            RpcController unused, TestProtos.EmptyRequestProto request)
+            throws ServiceException {
+      // Ensure clientId is received
+      byte[] clientId = Server.getClientId();
+      Assert.assertNotNull(clientId);
+      Assert.assertEquals(ClientId.BYTE_LENGTH, clientId.length);
+      // test with ip set
+      testSuccessLogFormat(true);
+      testFailureLogFormat(true);
+      return TestProtos.EmptyResponseProto.newBuilder().build();
+    }
+  }
+
+  /**
+   * Test {@link RouterAuditLogger} with IP set.
+   */
+  @Test
+  public void testRouterAuditLoggerWithIP() throws Exception {
+    Configuration conf = new Configuration();
+    RPC.setProtocolEngine(conf, TestRpcBase.TestRpcService.class, ProtobufRpcEngine2.class);
+
+    // Create server side implementation
+    MyTestRouterRPCServer serverImpl = new MyTestRouterRPCServer();
+    BlockingService service = TestRpcServiceProtos.TestProtobufRpcProto
+            .newReflectiveBlockingService(serverImpl);
+
+    // start the IPC server
+    Server server = new RPC.Builder(conf)
+            .setProtocol(TestRpcBase.TestRpcService.class)
+            .setInstance(service).setBindAddress("0.0.0.0")
+            .setPort(0).setNumHandlers(5).setVerbose(true).build();
+
+    server.start();
+
+    InetSocketAddress addr = NetUtils.getConnectAddress(server);
+
+    // Make a client connection and test the audit log
+    TestRpcBase.TestRpcService proxy = RPC.getProxy(TestRpcBase.TestRpcService.class,
+            TestRPC.TestProtocol.versionID, addr, conf);
+
+    // Start the testcase
+    TestProtos.EmptyRequestProto pingRequest =
+            TestProtos.EmptyRequestProto.newBuilder().build();
+    proxy.ping(null, pingRequest);
+
+    server.stop();
```

Comment:
This should be in a finally or similar to avoid leaking in case of failure.

Reply:
Thanks for your help reviewing the code, I will fix it.